---
title: STRING Database
category: papers
permalink: /STRING
---

# Summary

Paper:
[The STRING database in 2017: quality-controlled protein–protein association networks, made broadly accessible](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5210637/)

## Introduction

- Three main features:
    - Comprehensiveness
        - covers the largest number of organisms
        - uses the widest breadth of input sources (*automated text-mining and computational predictions*)
    - Usability
        - [intuitive web interface](https://string-db.org)
        - [Cytoscape](http://www.cytoscape.org) integration
        - [programmatic access options](http://string-db.org/help/api/)
    - Quality control & traceability
        - each interaction is annotated with benchmarked confidence scores, separately per evidence type, and the underlying evidence can be tracked to its source

## Database content

- Protein–protein association:
    - confidence score (i.e., the ‘edge weights’ in each network, scaled
to [0, 1]) <-> estimated likelihood that a given interaction is biologically meaningful, specific and reproducible, given the supporting evidence.

- Supporting evidence is divided into 7 channels:
    - assembled, scored and benchmarked separately
    - edges of different color in the web interface
    ![](https://string-db.org/api/image/network?identifiers=CCR5)
    - a combined into the confidence score for each interaction (final measure
for networks, sorting, filtering)


    - Main channels:
        - __Experiments__
            - actual experiments in the lab (biochemical, biophysical, genetic experiments).
            - populated mainly from the primary interaction databases (organized in the IMEx consortium, plus BioGRID)
        - __Database__
            - asserted by a human expert curator; this information is imported from pathway databases
        - __Textmining__
            - searching for mentions of protein names in all PubMed abstracts and in other text collections
            - pairs of proteins are given an association score when they are frequently mentioned together in the same paper, abstract or even sentence (relative to how often they are mentioned separately)
            - this score is raised further when it has been possible to parse one or more sentences through NLP, and a concept connecting the two proteins was encountered (such as ‘binding’ or ‘phosphorylation by’)
        - __Coexpression__
            - gene expression data originating from a variety of expression experiments are normalized, pruned and then correlated
            - pairs of proteins that are consistently similar in their expression patterns, under a variety of conditions, will receive a high association score
    

    - Genome-based prediction channels (functionality is generally most relevant for Bacteria and Archaea)
        - __Neighborhood__
            - genes are given an association score where they are consistently observed in each other’s genome neighborhood
        - __Fusion__
            - pairs of proteins are given an association score when there is at least one organism where their respective orthologs have fused into a single, protein-coding gene
        - __Co-occurrence__
            - evaluating the phylogenetic distribution of orthologs of all proteins in a given organism
            - if two proteins show a high similarity in this distribution, i.e. if their orthologs tend to be observed as ‘present’ or ‘absent’ in the same subsets of organisms, then an association score is assigned

- Using the transfer of evidence from one organism to another (‘interolog’ transfer):
    - based on the observation that orthologs of interacting proteins in one organism are often also interacting in another organism (better orthology relationships -> more confident inference)
    - all-against-all transfer of interactions, benchmarked separately for each evidence channel
    - transfers between closely related organisms are made more confidently, whereas the existence of paralogs (i.e., implied gene duplications) will lower the transfer score
    
## Web interface
- Exports:
    - as a bitmap image (PNG)
    - as a high-resolution bitmap (PNG, resolution at 400 dpi)
    - as a vector graphic (SVG)
    - as simple tabular text output (TSV)
    - as an XML summary (XML interaction data, according to the 'PSI-MI' data standard)
    - network coordinates: (flat-file format describing the coordinates and colors of nodes in the network)
    - protein sequences (MFA: multi-fasta format - containing the aminoacid sequences in the network)
    - protein annotations (a tab-delimited file describing the names, domains and annotated functions of the network proteins)

- Available downloads per query (TSV):
    - Biological Process
    - Molecular Function
    - Cellular Component
    - KEGG Pathways
    - PFAM Protein Domains
    - INTERPRO Protein Domains and Features

- Clustering (metric ?):
    - k-means
    - MCL

## Graph

- Edges of the graph (interactions, scored per evidence channel):
    - Evidence suggesting a functional link
    - Predictions for specific actions
    - Sequence similarities

- Nodes of the graph (actions):
    - Show protein sequence 
    - Homologs among STRING organisms 
    - Pathways, Functions, Resources (GeneCards)

## Download
- Available bigger data downloads for offline analysis: [Downloads](https://string-db.org/cgi/download.pl)

# Differences and our goal

STRING uses a [relational database scheme](https://stringdb-static.org/download/database.schema.v10.5.pdf), while we aim to construct the graph database. Having said that, graph databases allow more pratical inference on the data using machine learning algorithms.

Observed conceptual differences:
- STRING graph's edges do not contain direct semantic relation (e.g. _"binds to"_)
- Their text mining is based mainly on frequency of two proteins mentioned together in the same paper

In order to say what more we can offer than STRING and vice versa, we need to define the scope of the information we are looking for.

To start our project, we first focus on human (_Homo sapiens_) and mouse (_Mus musculus_).

Useful available properties for edges in the graph database:
- confidence score of an association
- predicted action

We could predict / approximate scores for _none / insignificant_ channels.
