# Structure of High IF Journals (Total number: 31)

This document contains the structure of Journals having Impact Factor greater than 15 - taken from [10x Genomics](https://www.10xgenomics.com/publications). Only relevant sections from the journals are added.

### Programming languages used
* R
* Python

### Source code found on
* Github
* Zenodo

### Sequencing Dataset found on
* European Genome-phenome Archive (EGA)
* Gene Expression Omnibus (GEO)
* Sequence Read Archive (SRA)
* International Nucleotide Sequence Database Collaboration (INSDC)
* European Nucleotide Archive (ENA)
* Genome Sequence Archive (GSA)
* BioStudies: ArrayExpress (Gene expression data)
* database of Genotypes and Phenotypes (dbGaP)
* ProteomeXchange Consortium - PRIDE: stores data related to protein and peptide identifications
* ProteomeXchange Consortium - PeptideAtlas: stores datasets focusing on peptide identifications and their associated spectral evidence
* ImmPort (Immunology Database and Analysis Portal)
* Zenodo
* China National GeneBank DataBase (CNGBdb)
* BioProject
* Medical Genomics Japan Variant Database (MGeND)
* CellxGene


_________________________________________________________________

## Cell: IF 45.5
_Structure 01:_
### Resource availability:
#### Materials availability
#### Data and Code/Software Availability
  * contains info regarding sequencing data and source code
  * if code unavailable > “This paper does not report original code” (MUST BE ADDED but not always added)
  * code available upon request from authors
  * already-existing SWs and Algorithms used but no unique code generated - found in Key Resources Table
  * ‘original code’
  * ‘unique code’
  * ‘custom code’
  * github link for original source code or custom code generated by using already-existing SW
 
_Structure 02:_
### Data and Code/Software Availability:
  * contains info regarding sequencing data and source code

_Structure 03:_ (in publications from 04/2011 to 2016)
### Accession Numbers
  * mostly written in plain text without a URL

## Cell Stem Cell: IF 19.8 
(some publications are not publicly accessible)
### Resource availability:
#### Lead contact
  * requests for resources and reagents
#### Materials availability 
#### Data and Code/Software Availability
  * section not found in several publications
  * sequencing data listed but not source code
  * 'original code'
  * 'novel code'
  * accession numbers of sequencing data & source code URL (Github/Zenodo) listed in Key Resources Table
  * accession number stated in plain text instead of hyperlinks/URLs

## Other Cell journals:
  * Cancer Cell: 48.8
  * Immunity: 25.5
  * Cell Host & Microbe: 20.6
  * Cell Metabolism: 27.7

## Nature Immunology: IF 27.7
_Structure 01:_
### Data availability 
  * uploaded on database(s) mentioned above
  * available upon request
  * fasta files found under Supplementary Information as Supplementary Data (observed in only one publication so far)
  * found under Source Data: ‘Fig.’ - .xlsx files

### Code availability
  * ‘custom code’
  * ‘custom software’
  * ‘new codes’
  * ‘customized code’
  * not provided
  * available from authors upon request
  * no new code, used existing code

_Structure 02:_
### Accession codes
  * contains URLs to sequencing data

**NOTE:** before 2019: Accession Codes, in 2019: Data Availability, 2020 onwards: Data Availability, Code Availability

## Other Nature journals:
  * Nature Biotechnology: 33.1
  * Nature Medicine: 58.7
  * Nature Genetics: 31.7
  * Nature: 50.5
  * Nature Cell Biology: 17.3
  * Nature Neuroscience: 21.2
  * Nature Cancer: 23.5
  * Nature Methods: 36.1
  * Nature Metabolism: 18.9
  * Nature Biomedical Engineering: 26.8
  * Nature Microbiology: 20.5
  * Cellular & Molecular Immunology: 21.8

## The Lancet Microbe: IF 20.9
### Data Sharing
  * contains accession numbers for sequencing data and source code URLs
  * accession numbers stated in plain text instead of hyperlink

## Other Lancet Journals:
  * The Lancet Oncology: 41.6
  * The Lancet Infectious Diseases: 36.4

## Cancer Discovery: IF 30.6
_Structure 01:_
### Data Availability/ Data Deposition/ Data and Software Availability/ Data and Materials Availability/ Data Reporting/ Code Available/ Data Archive (2019-present)
  * contains accession numbers for sequencing data and source code URLs
  * accession numbers stated in plain text instead of hyperlinks/URLs
  * very few publications contain source code

### Accession numbers
  * contains accession numbers of publicly available data used by authors

_Structure 02:_
### Methods:
#### Data Availability
  * information related to sequencing data
#### Software/Code Availability
  * source code

## Science Immunology: IF 19.1
### Data and materials availability
  * contains accession numbers for sequencing data and source code URLs

## Other Science Journals:
  * Science Translational Medicine: 16.9

## Nucleic Acids Research: IF 16.6
### Data Availability
  * contains information regarding sequencing data
  * few publications mention source code

## Cell Research: IF 28.2
### Data Availability/ Data Access
  * very few mentioned source code
  * accession numbers stated in plain text instead of hyperlink

## Gastroenterology: IF 25.7
### Data availability statement
  * contains accession codes and source code

## Journal of Hepatology: IF 26.8
### Data availability statement
  * contains accession codes (plain text) and source code (1 out of 19 publications)

## Gut: IF 23.0
### Data availability statement
  * contains accession codes (plain text) and source code (2 out of 13 publications)

## Circulation: IF 35.5
### Independent Data Access/ Data Availability/ Methods/ Data Resources
  * accession codes (plain text)
  * source code found (1 out of 23 publications)
  * available upon request


**General Observation:** Most publications started adding source code from 2019 onwards
