# Structure of High IF Journals

This document contains the structure of Journals having Impact Factor greater than 15 - taken from [10x Genomics](https://www.10xgenomics.com/publications). Only relevant sections from the journals are added.

### Programming languages used
* R
* Python

### Source code found on
* Github
* Zenodo

### Sequencing Dataset found on
* European Genome-phenome Archive (EGA)
* Gene Expression Omnibus (GEO)
* Sequence Read Archive (SRA)
* International Nucleotide Sequence Database Collaboration (INSDC)
* European Nucleotide Archive (ENA)
* Genome Sequence Archive (GSA)
* BioStudies: ArrayExpress (Gene expression data)
* database of Genotypes and Phenotypes (dbGaP)
* ProteomeXchange Consortium - PRIDE: stores data related to protein and peptide identifications
* ProteomeXchange Consortium - PeptideAtlas: stores datasets focusing on peptide identifications and their associated spectral evidence
* ImmPort (Immunology Database and Analysis Portal)
* Zenodo
* China National GeneBank DataBase (CNGBdb)
* BioProject


_________________________________________________________________

## Cell: IF 45.5
_Structure 01:_
### Resource availability:
#### Materials availability
#### Data and Code/Software Availability
  * contains info regarding sequencing data and source code
  * if code unavailable > “This paper does not report original code” (MUST BE ADDED but not always added)
  * code available upon request from authors
  * already-existing SWs and Algorithms used but no unique code generated - found in Key Resources Table
  * ‘original code’
  * ‘unique code’
  * ‘custom code’
  * github link for original source code or custom code generated by using already-existing SW
 
_Structure 02:_
### Data and Code/Software Availability:
  * contains info regarding sequencing data and source code

## Cell Stem Cell: IF 19.8 
(some publications are not publicly accessible)
### Resource availability:
#### Lead contact
  * requests for resources and reagents
#### Materials availability 
#### Data and Code/Software Availability
  * section not found in several publications
  * sequencing data listed but not source code
  * 'original code'
  * 'novel code'
  * accession numbers of sequencing data & source code URL (Github/Zenodo) listed in Key Resources Table
  * accession number stated in plain text instead of a URL

## Other Cell journals:
* Cancer Cell: 48.8
* Immunity: 25.5
* Cell Host & Microbe: 20.6
* Cell Metabolism: 27.7

## Nature Immunology: IF 27.7
### Data availability
  * uploaded on database(s) mentioned above
  * available upon request
  * fasta files found under Supplementary Information as Supplementary Data (observed in only one publication so far)
  * found under Source Data: ‘Fig.’ - .xlsx files

### Code availability
  * ‘custom code’
  * ‘custom software’
  * ‘new codes’
  * ‘customized code’
  * not provided
  * available from authors upon request
  * no new code, used existing code

## Other Nature journals:
* Nature Biotechnology: 33.1
* Nature Medicine: 58.7
* Nature Genetics: 31.7
* Nature: 50.5
* Nature Cell Biology: 17.3
* Nature Neuroscience: 21.2
* Nature Cancer: 23.5
* Nature Methods: 36.1
* Nature Metabolism: 18.9
* Nature Biomedical Engineering: 26.8
* Nature Microbiology: 20.5
* Nature Metabolism: 18.9


Note: Most of the Science journals cannot be accessed without Institution registration.
